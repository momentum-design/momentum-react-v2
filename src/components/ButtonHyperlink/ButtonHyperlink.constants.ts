const CLASS_PREFIX = 'md-button-hyperlink';

const DEFAULTS = {
  DISABLED: false,
  INVERTED: false,
};

const STYLE = {
  text: `${CLASS_PREFIX}-text`,
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
