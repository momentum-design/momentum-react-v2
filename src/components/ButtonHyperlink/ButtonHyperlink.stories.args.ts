import {
  commonAriaButton,
  commonHTMLAttributes,
  commonStyles,
} from '../../storybook/helper.stories.argtypes';

import { BUTTON_HYPERLINK_CONSTANTS as CONSTANTS } from './';

export default {
  ...commonStyles,
  ...commonHTMLAttributes,
  ...commonAriaButton,
  children: {
    description: 'Provides the child nodes for this element.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'ReactNode',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  disabled: {
    description: 'Whether to render the `<ButtonHyperlink />` as disabled.',
    options: [true, false],
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.DISABLED,
      },
    },
  },
  inverted: {
    description: 'Whether to render the `<ButtonHyperlink />` with inverted theme colors',
    options: [true, false],
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.INVERTED,
      },
    },
  },
};
