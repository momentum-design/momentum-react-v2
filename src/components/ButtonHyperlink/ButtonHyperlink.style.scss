.md-button-hyperlink-wrapper {
  background-color: rgba(0, 0, 0, 0);
  border: var(--md-globals-border-clear);
  color: var(--mds-color-theme-text-accent-normal);
  cursor: pointer;
  font-family: var(--md-globals-font-default);
  outline: none !important;
  padding: 0;
  text-decoration: underline;
  transition: background-color 0.2s, color 0.2s, border-color 0.2s;

  &:focus,
  &.focus {
    color: var(--mds-color-theme-text-accent-normal);
  }

  &:hover,
  &.hover {
    color: var(--mds-color-theme-text-accent-hover);
  }

  &:active,
  &.active {
    color: var(--mds-color-theme-text-accent-active);
  }

  &[data-inverted='true'] {
    color: var(--mds-color-theme-inverted-text-accent-normal);

    &:focus,
    &.focus {
      color: var(--mds-color-theme-inverted-text-accent-normal);
    }

    &:hover,
    &.hover {
      color: var(--mds-color-theme-inverted-text-accent-hover);
    }

    &:active,
    &.active {
      color: var(--mds-color-theme-inverted-text-accent-hover);
    }

    &[data-disabled='true'],
    &.disable {
      color: var(--mds-color-theme-text-primary-disabled);
      cursor: auto;
    }
  }

  &[data-disabled='true'],
  &.disable {
    color: var(--mds-color-theme-text-primary-disabled);
    cursor: auto;
  }
}
