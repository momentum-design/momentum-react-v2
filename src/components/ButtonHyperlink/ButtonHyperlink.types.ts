import { ReactNode } from 'react';
import { AriaButtonProps } from '@react-types/button';

export interface Props extends AriaButtonProps<'a'> {
  /**
   * Custom class to be able to override the component's CSS
   */
  className?: string;
  /**
   * Child components of this ButtonDialpad. This prop replaces primaryText.
   */
  children?: ReactNode;

  /**
   * Whether or not this ButtonDialpad is disabled.
   */
  disabled?: boolean;

  /**
   * Whether or not this ButtonDialpad has inverted theme colors.
   */
  inverted?: boolean;

  /**
   * title to use for this component.
   */
  title?: string;
}
