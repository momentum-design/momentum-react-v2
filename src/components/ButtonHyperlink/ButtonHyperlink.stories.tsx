import { MultiTemplateWithPseudoStates, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';
import AriaButtonDocs from '../../storybook/docs.stories.aria-button.mdx';
import HTMLGlobalAttributes from '../../storybook/docs.stories.common-html-attributes.mdx';

import ButtonHyperlink, { ButtonHyperlinkProps } from './';
import argTypes from './ButtonHyperlink.stories.args';
import Documentation from './ButtonHyperlink.stories.docs.mdx';

export default {
  title: 'Momentum UI/ButtonHyperlink',
  component: ButtonHyperlink,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs, HTMLGlobalAttributes, AriaButtonDocs),
    },
  },
  args: {
    children: 'Example Text',
  },
};

const Example = Template<ButtonHyperlinkProps>(ButtonHyperlink).bind({});

Example.argTypes = { ...argTypes };

const States = MultiTemplateWithPseudoStates<ButtonHyperlinkProps>(ButtonHyperlink).bind({});

States.parameters = {
  variants: [{ label: 'Default hyperlink' }, { inverted: true }],
};

States.argTypes = { ...argTypes };
delete States.argTypes.disabled;
delete States.argTypes.inverted;

export { Example, States };
