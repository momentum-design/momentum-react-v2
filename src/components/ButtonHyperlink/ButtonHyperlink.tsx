import React, { forwardRef, useRef, RefObject } from 'react';
import { useButton } from '@react-aria/button';
import FocusRing from '../FocusRing';

import { DEFAULTS, STYLE } from './ButtonHyperlink.constants';
import { Props } from './ButtonHyperlink.types';
import './ButtonHyperlink.style.scss';
import classnames from 'classnames';

const ButtonHyperlink = forwardRef((props: Props, providedRef: RefObject<HTMLAnchorElement>) => {
  const { className, title } = props;
  const internalRef = useRef();
  const ref = providedRef || internalRef;

  const mutatedProps = {
    ...props,
    isDisabled: props.disabled,
    isInverted: props.inverted,
  };

  delete mutatedProps.disabled;
  delete mutatedProps.className;

  const { buttonProps } = useButton({ ...mutatedProps, elementType: 'a' }, ref);

  return (
    <FocusRing disabled={props.disabled}>
      <a
        className={classnames(STYLE.wrapper, className)}
        {...buttonProps}
        ref={ref}
        data-disabled={props.disabled || DEFAULTS.DISABLED}
        data-inverted={props.inverted || DEFAULTS.INVERTED}
        title={title}
      >
        {props.children}
      </a>
    </FocusRing>
  );
});

ButtonHyperlink.displayName = 'ButtonHyperlink';

export default ButtonHyperlink;
