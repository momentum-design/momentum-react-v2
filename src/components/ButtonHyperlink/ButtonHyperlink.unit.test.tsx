import React from 'react';
import { mount } from 'enzyme';
import { ButtonHyperlink } from '@momentum-ui/react-collaboration';

import { DEFAULTS, STYLE } from './ButtonHyperlink.constants';

describe('ButtonHyperlink', () => {
  describe('snapshot', () => {
    let container;

    it('should match snapshot', () => {
      expect.assertions(1);

      container = mount(<ButtonHyperlink>Hyperlink</ButtonHyperlink>);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot when disabled', () => {
      expect.assertions(1);

      const disabled = !DEFAULTS.DISABLED;

      container = mount(<ButtonHyperlink disabled={disabled}>Hyperlink</ButtonHyperlink>);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot when inverted', () => {
      expect.assertions(1);

      const inverted = !DEFAULTS.INVERTED;

      container = mount(<ButtonHyperlink inverted={inverted}>Hyperlink</ButtonHyperlink>);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with title', () => {
      expect.assertions(1);

      const title = 'Example Text';

      const container = mount(<ButtonHyperlink title={title} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its main class', () => {
      expect.assertions(1);

      const element = mount(<ButtonHyperlink />)
        .find(ButtonHyperlink)
        .getDOMNode();

      expect(element.classList.contains(STYLE.wrapper)).toBe(true);
    });

    it('should have custom class if provided', () => {
      const testClass = 'testClass';

      const wrapper = mount(<ButtonHyperlink className={testClass} />);
      const element = wrapper.find(ButtonHyperlink).getDOMNode();

      expect(element.classList.contains(testClass)).toBe(true);
    });

    it('should pass disabled prop', () => {
      expect.assertions(1);

      const disabled = !DEFAULTS.DISABLED;

      const element = mount(<ButtonHyperlink disabled={disabled} />)
        .find(ButtonHyperlink)
        .getDOMNode();

      expect(element.getAttribute('data-disabled')).toBe(`${disabled}`);
    });

    it('should pass inverted prop', () => {
      expect.assertions(1);

      const inverted = !DEFAULTS.INVERTED;

      const element = mount(<ButtonHyperlink inverted={inverted} />)
        .find(ButtonHyperlink)
        .getDOMNode();

      expect(element.getAttribute('data-inverted')).toBe(`${inverted}`);
    });

    it('should pass child prop', () => {
      expect.assertions(1);

      const child = '1';

      const element = mount(<ButtonHyperlink>{child}</ButtonHyperlink>)
        .find(ButtonHyperlink)
        .getDOMNode();

      expect(element.innerHTML).toBe(child);
    });

    it('should have provided title when title is provided', () => {
      expect.assertions(1);

      const title = 'Example Text';

      const element = mount(<ButtonHyperlink title={title} />)
        .find(ButtonHyperlink)
        .getDOMNode();

      expect(element.getAttribute('title')).toBe(title);
    });
  });

  describe('actions', () => {
    it('should handle mouse press events', () => {
      expect.assertions(1);

      const mockCallback = jest.fn();

      const component = mount(<ButtonHyperlink onPress={mockCallback} />).find(ButtonHyperlink);

      component.props().onPress({
        type: 'press',
        pointerType: 'mouse',
        altKey: false,
        shiftKey: false,
        ctrlKey: false,
        metaKey: false,
        target: component.getDOMNode(),
      });

      expect(mockCallback).toBeCalledTimes(1);
    });
  });
});
