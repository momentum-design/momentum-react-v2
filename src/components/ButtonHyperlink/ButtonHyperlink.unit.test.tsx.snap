// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonHyperlink snapshot should match snapshot 1`] = `
<ButtonHyperlink>
  <FocusRing>
    <FocusRing
      focusClass="md-focus-ring-wrapper children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <a
        className="md-button-hyperlink-wrapper"
        data-disabled={false}
        data-inverted={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        Hyperlink
      </a>
    </FocusRing>
  </FocusRing>
</ButtonHyperlink>
`;

exports[`ButtonHyperlink snapshot should match snapshot when disabled 1`] = `
<ButtonHyperlink
  disabled={true}
>
  <FocusRing
    disabled={true}
  >
    <FocusRing
      disabled={true}
      focusClass="md-focus-ring-wrapper-disabled"
      focusRingClass="md-focus-ring-wrapper-disabled"
    >
      <a
        aria-disabled={true}
        className="md-button-hyperlink-wrapper"
        data-disabled={true}
        data-inverted={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
      >
        Hyperlink
      </a>
    </FocusRing>
  </FocusRing>
</ButtonHyperlink>
`;

exports[`ButtonHyperlink snapshot should match snapshot when inverted 1`] = `
<ButtonHyperlink
  inverted={true}
>
  <FocusRing>
    <FocusRing
      focusClass="md-focus-ring-wrapper children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <a
        className="md-button-hyperlink-wrapper"
        data-disabled={false}
        data-inverted={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        Hyperlink
      </a>
    </FocusRing>
  </FocusRing>
</ButtonHyperlink>
`;

exports[`ButtonHyperlink snapshot should match snapshot with title 1`] = `
<ButtonHyperlink
  title="Example Text"
>
  <FocusRing>
    <FocusRing
      focusClass="md-focus-ring-wrapper children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <a
        className="md-button-hyperlink-wrapper"
        data-disabled={false}
        data-inverted={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
        title="Example Text"
      />
    </FocusRing>
  </FocusRing>
</ButtonHyperlink>
`;
